JSON Rest APIs

katonaaron-msg opened the pull request:
- Add product service, mapper, controller and DTOs.

katonaaron-msg committed:
Add product service, mapper, controller and DTOs.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/mapper/ProductMapper.java:
```diff
@@ -0,0 +1,46 @@
+public abstract class ProductMapper {
+
+    @Autowired
+    protected ProductCategoryRepository categoryRepository;
```
Do not use repositories in a mapper class. A Mapper should just map data from an entity to a Dto and from Dto to entity.  All the other logic which implies repositories should live outside the mapper, in the service class eventually.

katonaaron-msg committed:
Change ProductMapper to not use the repositories.

katonaaron-msg committed:
Implement order creation

Implement OrderService, OrderController and the necessary DTOs and mappers.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/service/OrderService.java:
```diff
@@ -0,0 +1,75 @@
+    private final AddressMapper addressMapper;
+
+    @Transactional
+    public Order createOrder(CreateOrderDTO createOrderDTO) {
```
Here you either use Order as a parameter, or you return an OrderDto. But it is not ok for the method to receive a Dto and return an entity (in this case Order). Usually, services work with entities and the controller is the one who does the mapping to Dto. Or there is another option in which the service receives Dto and returns again a Dto and the mapping to the domain is done in the methods inside the service. But there is never a mix between the two options. The preferred option is option 1, where the services works with entities and the controller performs the mapping between entity and dto.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/service/OrderService.java:
```diff
@@ -0,0 +1,75 @@
+    public Order createOrder(CreateOrderDTO createOrderDTO) {
+        final var productIdToQuantity = createOrderDTO.orderDetails().stream()
+                .collect(Collectors.toMap(CreateOrderDetailDTO::productId, CreateOrderDetailDTO::quantity));
+        final var location = locationRepository.findAll().stream()
```
Can you write a query for this complex operation? It is very expensive to retrieve all the locations first and then filter the wrong ones out. If you have hundreds of locations and only one of these locations is the correct one, you will load all the locations in memory which is an expensive operation.

cristinaAnca commented on src/main/java/ro/msg/learning/shop/service/OrderService.java:
```diff
@@ -0,0 +1,75 @@
+        order.setOrderDetails(orderDetails);
+        order.setShippedFrom(location);
+        order.setDeliveryAddress(addressMapper.toEntity(createOrderDTO.deliveryAddress()));
+        order.setCustomer(null); //TODO
```
You need to solve the TODO comment, customer cannot be null

katonaaron-msg committed:
Rewrite order creation

- Created a repository method for finding the locations where a product is available in a given quantity. Intersected these locations for each product.
- Created a dummy customer which is used for each order.
- Changed the method parameters to eliminate the dto

katonaaron-msg committed:
Select a potential location based on the smallest id

cristinaAnca commented on src/main/java/ro/msg/learning/shop/ShopApplication.java:
```diff
@@ -2,12 +2,35 @@
+	 * @return
+	 */
+	@Bean
+	Customer customer(CustomerRepository customerRepository) {
```
I understand what you want to do with this bean, and it's a good idea. But let's move it elsewhere, in a configuration file, just to be separated from the Application. You have more details here: https://docs.spring.io/spring-javaconfig/docs/1.0.0.m3/reference/html/creating-bean-definitions.html.

katonaaron-msg replied:
I moved it to a separate configuration file :)

katonaaron-msg committed:
Move dummy customer to a separate configuration file
